Add List All Model Versions Query

StephenODea54 opened the pull request:
Ref #54 

Created a `listAllModelVersions` query that returns all registered models with their most recent model version. I added a new function instead of altering the `listModelVersions` function because that method is used for returning all model versions for a particular model.

I am unable to test this out, as the database schema isn't being created when I run `bootstrap.sh` in the `apollo-infra` directory. Nonetheless, it's a pretty simple method so I think it's good to go. Once I figure out what is going on with the database I'll be able to come back and test.

StephenODea54 committed:
Add ListAllModelVersions Query

StephenODea54 committed:
Add list all model versions query on frontend

wetherc commented:
You have to `./bin/storage upgrade` after bootstrapping for the patches to be applied

wetherc commented on dstk-infra/apollo/src/graphql/model-version/modelVersionQueries.ts:
```diff
@@ -16,3 +16,20 @@ export const ListModelVersions = extendType({
+                const result = await ObjectionMLModelVersion.query().where(
+                    'numericVersion',
+                    '=',
+                    ObjectionMLModelVersion.query().max('numericVersion'),
```
Won't this give you versions equal to the global maximum, instead of the max within a given model?

StephenODea54 replied:
Oh yeah, good catch. Think I need to add a `.groupBy('modelId')`

wetherc replied:
I'm not sure that would solve this. I think that whatever solution you come to will end up having something like a subquery with a RANK or similar; there's not (unless I'm having a major brain fart which, I mean, is always a legit possiblity) a straightfoward way to do this operation without the SQL getting a little weird. (cf. https://stackoverflow.com/a/2657495)

If you want to make life a little easier on yourself, we could I think take one of 2 routes:

  1. Just run a simple query for each model to get its max model version. This has the downside of being many small queries.
  2. Add a `currentModelVersion` column to `registry.models` that's a FK reference to whatever the most recent version for that model is at the time. This would require some small edits to the `CreateModelVersion` mutation, but that's like a 2-line change. From there, you could just call it a day with

```typescript
const result = await ObjectionMLModel.relatedQuery('currentModelVersion') as ObjectionMLModelVersion;
```

StephenODea54 replied:
I'll get back to you on this. I'm getting a `readarray: command not found` when trying to apply the postgres patches. I don't really know about shells and what not, so I'm going to procrastinate by getiing my Emergency Domino's pizza and looking at Christmas lights 😂

I'll give this another look tonight or tomorrow morning.

wetherc replied:
Ohhhh probably your shell is zsh and not bash; I forgot `readarray` doesn't port over to zsh cleanly. It should have picked up the hash bang at the top of the file though... Weird

In the meanwhile, try `bash ./bin/storage upgrade`

StephenODea54 replied:
Yeah I tried that already still not recognizing the `readarray` command. I'm going to try upgrading bash and see if that works.

StephenODea54 replied:
Upgrading bash worked!

wetherc replied:
Ahh yeah Apple ships computers with older versions of bash, that could be it. #74 should add ZSH compatibility, which is the recommended shell for macs nowadays, so that's also an option if you just want to make that switch locally

StephenODea54 commented:
> You have to `./bin/storage upgrade` after bootstrapping for the patches to be applied

🙃 duh

wetherc committed:
Encrypt storage provider secrets (#72)

Resolves #26

I hate cryptography and existence is pain. This adds encryption
to storage provider secrets using ChaCha20-Poly1305. Decryption
of the Access Key ID is automatic when returning the value to
a GQL client, and the secret access key is never displayed to
the client again after being entered, although will be accessed
and decrypted as needed to perform operations against the upstream
storage provider.

There's an outstanding issue with adding the secret key as an
environment variable in the k8s manifest that I haven't resolved
yet, so I'm punting on it for the moment and have a temporary and

    IMPORTANT: TOTALLY INSECURE (intentionally)

key hardcoded in the encryption utility itself. Obviously this is
A Bad Thing (tm) and is just a stopgap until I get my act together.

wetherc committed:
ZSH cross-shell compatibility for storage script (#74)

wetherc committed:
Port Nexus to Pothos (#76)

* Nexus->Pothos port

* working storage provider methods

* ported model methods

* model version ports

* presigned URL port

* clean up an eslint comment

StephenODea54 committed:
Add Current Model Version Patch

StephenODea54 committed:
Add updated current model version query to frontend

StephenODea54 committed:
Merge branch 'dev' into add-list-model-query-frontend

wetherc commented on dstk-infra/apollo/src/graphql/model/model.ts:
```diff
@@ -1,36 +1,42 @@
             },
         }),
 
-        isArchived: t.exposeBoolean('isArchived'),
```
I'd prefer we not delete this. Pothos makes an explicit choice to have this abstraction layer over top of the underlying data model, and I'd generally like to stay idiomatic wrt package conventions

StephenODea54 replied:
Oh sorry, that must have been a merge conflict mishap. Fixed.

wetherc commented on dstk-infra/postgres/patches/20231226_current-model-version-key.sql:
```diff
@@ -0,0 +1,7 @@
+\connect dstk;
+
+CREATE EXTENSION IF NOT EXISTS "uuid-ossp";
```
This isn't needed

StephenODea54 replied:
Just the `CREATE EXTENSION...` bit?

wetherc replied:
Yeah, that's an only-needs-to-happen-once but that gets taken care of when the db is very first created. It gets ignored on subsequent runs, so no need to have it in here

StephenODea54 replied:
Makes sense, thank you for the explanation!

StephenODea54 committed:
Fix MLModel object type

StephenODea54 committed:
Remove UUID extension in currentModelVersion patch
